_posts/2016-01-26-variance-and-functors.md

```diff
@@ -0,0 +1,309 @@
+
+## Array
+A type that cannot safely be made covariant is `Array`. If `Array` were covariant, we could substitute
+an `Array[Circle]` for an `Array[Shape]`. This can get us in a nasty situation.
+
+```scala
+val circles: Array[Circle] = Array.fill(10)(Circle(..))
+val shapes: Array[Shape] = circles // works only if Array is covariant
+shapes(0) = Square(..) // Square is a subtype of Shape
+```
+
+If `Array` was covariant this would compile fine, but fail at runtime. In fact, Java arrays are
+covariant and so the analogous Java code would compile, throwing an `ArrayStoreException` when
+run. The compiler accepts this because it is valid to upcast an `Array[Circle]` into an `Array[Shape]`,
+and it is valid to insert a `Shape` into an `Array[Shape]`. However the runtime representation of
+`shapes` is still an `Array[Circle]` and inserting a `Square` into it isn't allowed.
+
+## Read-only and covariance
+In general, a type can be made safely covariant if it is read-only. If we know how to read a specific type, we know
+how to read a more general type by throwing away any extra information. `List` is safe to to make
+covariant because it is immutable and we can only ever read information off of it. With `Array`, we
+cannot make it covariant because we are able to write to it.
+
+## Functor
+We can also encode the covariance behavior in an explicit interface, commonly called `Functor`.
+
+```scala
+trait Functor[F[_]] {
+  def map[A, B](fa: F[A])(f: A => B): F[B]
+}
+```
+
+We can implement `Functor` for `List` and `Read`.
+
+```scala
+val listFunctor: Functor[List] =
+  new Functor[List] {
+    def map[A, B](fa: List[A])(f: A => B): List[B] = fa match {
+      case Nil => Nil
+      case a :: as => f(a) :: map(as)(f)
+    }
+  }
+
+val readFunctor: Functor[Read] =
+  new Functor[Read] {
+    def map[A, B](fa: Read[A])(f: A => B): Read[B] =
+      new Read[B] {
+        def read(s: String): Option[B] =
+          fa.read(s) match {
+            case None => None
+            case Some(a) => Some(f(a))
+          }
+      }
+  }
+```
+
+With that we can do useful things like
+
+```scala
+val circles: List[Circle] = List(Circle(..), Circle(..))
+val shapes: List[Shape] = listFunctor.map(circles)(circle => circle: Shape) // upcast
+
+val parseCircle: Read[Circle] = ...
+val parseShape: Read[Shape] = readFunctor.map(parseCircle)(circle => circle: Shape) // upcast
+```
+
+or more generally:
+
+```scala
+def upcast[F[_], A, B <: A](functor: Functor[F], fb: F[B]): F[A] =
+  functor.map(fb)(b => b: A)
+```
+
+`upcast`'s behavior does exactly what covariance does - given some supertype `A` (`Shape`) and a subtype `B` (`Circle`),
+we can mechanically (and safely) turn an `F[B]` into an `F[A]`. Put differently, anywhere we expect an `F[A]` we can provide
+an `F[B]`, i.e. covariance. For this reason, `Functor` is sometimes referred to in full as covariant functor.
+
+# Contravariance
```

Comment:
You might suggest that contravariance (of either sort) most often applies for datatypes that represent computations; this is because the only primitive-contravariant position is the fucntion argument. For example, the input to a Kleisli arrow, and [the "haystack" (`S`) to a fold](https://github.com/julien-truffaut/Monocle/blob/8efd6c85f8a3697abc11feae0a5f1a2ba7fd0a58/core/src/main/scala/monocle/Fold.scala#L28), are contravariant.